### ISSUE TITLE
Pyright 'forgets' type of instance variable in method when type is Literal

### ISSUE THREAD

**Issue by couteau** [microsoft/pyright#10448]
When I assign an instance variable a value that is typed as a Literal type, pyright seems to forget that type when the instance variable is later used in a method on the instance. Example below:

```python
from typing import Literal, reveal_type

type MyType = Literal['a', 'b']


my_dict: dict[MyType, str] = {'a': 'foo', 'b': 'bar'}

class Test:
    def __init__(self, t: MyType):
        self.t = t
        reveal_type(self.t)

    def test(self) -> str:
        reveal_type(self.t)
        return my_dict[self.t]
```

when running pyright from the command-line, the output is:
```
test.py:11:21 - information: Type of "self.t" is "Literal['a', 'b']"
test.py:14:21 - information: Type of "self.t" is "str"
test.py:15:16 - error: Argument of type "str" cannot be assigned to parameter "key" of type "MyType" in function "__getitem__"
    Type "str" is not assignable to type "MyType"
      "str" is not assignable to type "Literal['a']"
      "str" is not assignable to type "Literal['b']" (reportArgumentType)
1 error, 0 warnings, 2 informations
```
I see the same error even if I explicitly type hint `self.t` to be `MyType`. Pyright should recognize that `self.t` is `Literal['a', 'b']` inside the `test` method. MyPy correctly reports the type of `self.t` in both places.  

I'm using pyright version 1.1.400 installed via anaconda on MacOS Sequoia.

**Comment by erictraut** [comment]
Pyright is working as designed here, so this isn't a bug.

In your code sample, you haven't provided a type annotation for the instance variable `t`, so its type is inferred based on all assignments within the class implementation. Literals are not retained during this inference. For example, if you assigned `self.t = 1`, the right-hand side evaluates to `Literal[1]`, but pyright infers the type of `t` to be `int`. 

If you want the instance variable `t` to be limited to a literal value (or union of literal values) you need to declare it as such.

```python
self.t: MyType = t
```
